<html>
<head>
<title>Python/errors.c</title>
<meta name='robots' content='noindex,nofollow'>
<meta name='generator' content='GLOBAL-5.8.1'>
</head>
<body text='#191970' bgcolor='#f5f5dc' vlink='gray'>
<a name='TOP'><h2><a href='../mains.html'>root</a>/<a href='../files/3402.html'>Python</a>/errors.c</h2>
<i><font color='green'>/* [&lt;][&gt;]<a href='#L25'>[^]</a><a href='#L848'>[v]</a>[top]<a href='#BOTTOM'>[bottom]</a><a href='../mains.html'>[index]</a><a href='../help.html'>[help]</a> */</font></i>
<hr>
<h2>DEFINITIONS</h2>
This source file includes following definitions.
<ol>
<li><a href='#L25' title='Defined at 25.'>PyErr_Restore</a>
<li><a href='#L53' title='Defined at 53.'>PyErr_SetObject</a>
<li><a href='#L116' title='Defined at 116.'>PyErr_SetNone</a>
<li><a href='#L122' title='Defined at 122.'>PyErr_SetString</a>
<li><a href='#L131' title='Defined at 131.'>PyErr_Occurred</a>
<li><a href='#L145' title='Defined at 145.'>PyErr_GivenExceptionMatches</a>
<li><a href='#L189' title='Defined at 189.'>PyErr_ExceptionMatches</a>
<li><a href='#L202' title='Defined at 202.'>PyErr_NormalizeException</a>
<li><a href='#L304' title='Defined at 304.'>PyErr_Fetch</a>
<li><a href='#L318' title='Defined at 318.'>PyErr_Clear</a>
<li><a href='#L326' title='Defined at 326.'>PyErr_BadArgument</a>
<li><a href='#L334' title='Defined at 334.'>PyErr_NoMemory</a>
<li><a href='#L341' title='Defined at 341.'>PyErr_SetFromErrnoWithFilenameObject</a>
<li><a href='#L430' title='Defined at 430.'>PyErr_SetFromErrnoWithFilename</a>
<li><a href='#L440' title='Defined at 440.'>PyErr_SetFromErrnoWithUnicodeFilename</a>
<li><a href='#L452' title='Defined at 452.'>PyErr_SetFromErrno</a>
<li><a href='#L459' title='Defined at 459.'>PyErr_SetExcFromWindowsErrWithFilenameObject</a>
<li><a href='#L513' title='Defined at 513.'>PyErr_SetExcFromWindowsErrWithFilename</a>
<li><a href='#L526' title='Defined at 526.'>PyErr_SetExcFromWindowsErrWithUnicodeFilename</a>
<li><a href='#L541' title='Defined at 541.'>PyErr_SetExcFromWindowsErr</a>
<li><a href='#L546' title='Defined at 546.'>PyErr_SetFromWindowsErr</a>
<li><a href='#L551' title='Defined at 551.'>PyErr_SetFromWindowsErrWithFilename</a>
<li><a href='#L563' title='Defined at 563.'>PyErr_SetFromWindowsErrWithUnicodeFilename</a>
<li><a href='#L579' title='Defined at 579.'>_PyErr_BadInternalCall</a>
<li><a href='#L590' title='Defined at 590.'>PyErr_BadInternalCall</a>
<li><a href='#L600' title='Defined at 600.'>PyErr_Format</a>
<li><a href='#L621' title='Defined at 621.'>PyErr_NewException</a>
<li><a href='#L673' title='Defined at 673.'>PyErr_NewExceptionWithDoc</a>
<li><a href='#L708' title='Defined at 708.'>PyErr_WriteUnraisable</a>
<li><a href='#L764' title='Defined at 764.'>PyErr_SyntaxLocation</a>
<li><a href='#L774' title='Defined at 774.'>PyErr_SyntaxLocationEx</a>
<li><a href='#L848' title='Defined at 848.'>PyErr_ProgramText</a>
</ol>
<hr>
<pre>
<a name='L1'>
<a name='L2'><i><font color='green'>/* Error handling */</font></i>
<a name='L3'>
<a name='L4'><font color='darkred'>#include</font> "<a href='570.html'>Python.h</a>"
<a name='L5'>
<a name='L6'><font color='darkred'>#ifndef</font> __STDC__
<a name='L7'><font color='darkred'>#ifndef</font> <a href='../S/2906.html#L81' title='Defined at 81 in PC/pyconfig.h.'>MS_WINDOWS</a>
<a name='L8'><b>extern</b> <b>char</b> *strerror(<b>int</b>);
<a name='L9'><font color='darkred'>#endif</font>
<a name='L10'><font color='darkred'>#endif</font>
<a name='L11'>
<a name='L12'><font color='darkred'>#ifdef</font> <a href='../S/2906.html#L81' title='Defined at 81 in PC/pyconfig.h.'>MS_WINDOWS</a>
<a name='L13'><font color='darkred'>#include</font> &lt;windows.h&gt;
<a name='L14'><font color='darkred'>#include</font> &lt;winbase.h&gt;
<a name='L15'><font color='darkred'>#endif</font>
<a name='L16'>
<a name='L17'><font color='darkred'>#include</font> &lt;<a href='2809.html'>ctype.h</a>&gt;
<a name='L18'>
<a name='L19'><font color='darkred'>#ifdef</font> __cplusplus
<a name='L20'><b>extern</b> "C" <font color='red'>{</font>
<a name='L21'><font color='darkred'>#endif</font>
<a name='L22'>
<a name='L23'>
<a name='L24'><b>void</b>
<a name='L25'><a href='../R/2881.html' title='Multiple refered from 36 places.'>PyErr_Restore</a>(PyObject *type, PyObject *value, PyObject *traceback)
<a name='L26'><font color='red'>{</font>
<a name='L27'>    <a href='../D/3832.html' title='Multiple defined in 2 places.'>PyThreadState</a> *tstate = <a href='../S/509.html#L90' title='Defined at 90 in Include/ceval.h.'>PyThreadState_GET</a>();
<a name='L28'>    <a href='../D/3593.html' title='Multiple defined in 2 places.'>PyObject</a> *oldtype, *oldvalue, *oldtraceback;
<a name='L29'>
<a name='L30'>    <b>if</b> (traceback != <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a> &amp;&amp; !PyTraceBack_Check(traceback)) <font color='red'>{</font>
<a name='L31'>        <i><font color='green'>/* XXX Should never happen -- fatal error instead? */</font></i>
<a name='L32'>        <i><font color='green'>/* Well, it could be None. */</font></i>
<a name='L33'>        Py_DECREF(traceback);
<a name='L34'>        traceback = <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>;
<a name='L35'>    <font color='red'>}</font>
<a name='L36'>
<a name='L37'>    <i><font color='green'>/* Save these in locals to safeguard against recursive</font></i>
<a name='L38'><i><font color='green'>       invocation through Py_XDECREF */</font></i>
<a name='L39'>    oldtype = tstate-&gt;curexc_type;
<a name='L40'>    oldvalue = tstate-&gt;curexc_value;
<a name='L41'>    oldtraceback = tstate-&gt;curexc_traceback;
<a name='L42'>
<a name='L43'>    tstate-&gt;curexc_type = type;
<a name='L44'>    tstate-&gt;curexc_value = <a href='../S/2252.html#L27' title='Defined at 27 in Modules/_ctypes/ctypes.h.'>value</a>;
<a name='L45'>    tstate-&gt;curexc_traceback = traceback;
<a name='L46'>
<a name='L47'>    <a href='../S/544.html#L767' title='Defined at 767 in Include/object.h.'>Py_XDECREF</a>(oldtype);
<a name='L48'>    <a href='../S/544.html#L767' title='Defined at 767 in Include/object.h.'>Py_XDECREF</a>(oldvalue);
<a name='L49'>    <a href='../S/544.html#L767' title='Defined at 767 in Include/object.h.'>Py_XDECREF</a>(oldtraceback);
<a name='L50'><font color='red'>}</font>
<a name='L51'>
<a name='L52'><b>void</b>
<a name='L53'><a href='../R/2895.html' title='Multiple refered from 61 places.'>PyErr_SetObject</a>(PyObject *exception, PyObject *value)
<a name='L54'><font color='red'>{</font>
<a name='L55'>    <a href='../D/3832.html' title='Multiple defined in 2 places.'>PyThreadState</a> *tstate = <a href='../S/509.html#L90' title='Defined at 90 in Include/ceval.h.'>PyThreadState_GET</a>();
<a name='L56'>    <a href='../D/3593.html' title='Multiple defined in 2 places.'>PyObject</a> *exc_value;
<a name='L57'>    <a href='../D/3593.html' title='Multiple defined in 2 places.'>PyObject</a> *tb = <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>;
<a name='L58'>
<a name='L59'>    <b>if</b> (exception != <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a> &amp;&amp;
<a name='L60'>        !PyExceptionClass_Check(exception)) <font color='red'>{</font>
<a name='L61'>        <a href='../S/3106.html#L600' title='Defined at 600 in Python/errors.c.'>PyErr_Format</a>(PyExc_SystemError,
<a name='L62'>                     "exception %R not a BaseException subclass",
<a name='L63'>                     exception);
<a name='L64'>        <b>return</b>;
<a name='L65'>    <font color='red'>}</font>
<a name='L66'>    <a href='../S/544.html#L766' title='Defined at 766 in Include/object.h.'>Py_XINCREF</a>(<a href='../S/2252.html#L27' title='Defined at 27 in Modules/_ctypes/ctypes.h.'>value</a>);
<a name='L67'>    exc_value = tstate-&gt;exc_value;
<a name='L68'>    <b>if</b> (exc_value != <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a> &amp;&amp; exc_value != Py_None) <font color='red'>{</font>
<a name='L69'>        <i><font color='green'>/* Implicit exception chaining */</font></i>
<a name='L70'>        Py_INCREF(exc_value);
<a name='L71'>        <b>if</b> (<a href='../S/2252.html#L27' title='Defined at 27 in Modules/_ctypes/ctypes.h.'>value</a> == <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a> || !PyExceptionInstance_Check(<a href='../S/2252.html#L27' title='Defined at 27 in Modules/_ctypes/ctypes.h.'>value</a>)) <font color='red'>{</font>
<a name='L72'>            <i><font color='green'>/* We must normalize the value right now */</font></i>
<a name='L73'>            <a href='../D/3593.html' title='Multiple defined in 2 places.'>PyObject</a> *<a href='../S/2858.html#L294' title='Defined at 294 in PC/bdist_wininst/install.c.'>args</a>, *fixed_value;
<a name='L74'>            <b>if</b> (<a href='../S/2252.html#L27' title='Defined at 27 in Modules/_ctypes/ctypes.h.'>value</a> == <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a> || <a href='../S/2252.html#L27' title='Defined at 27 in Modules/_ctypes/ctypes.h.'>value</a> == Py_None)
<a name='L75'>                <a href='../S/2858.html#L294' title='Defined at 294 in PC/bdist_wininst/install.c.'>args</a> = <a href='../S/2823.html#L49' title='Defined at 49 in Objects/tupleobject.c.'>PyTuple_New</a>(0);
<a name='L76'>            <b>else</b> <b>if</b> (PyTuple_Check(<a href='../S/2252.html#L27' title='Defined at 27 in Modules/_ctypes/ctypes.h.'>value</a>)) <font color='red'>{</font>
<a name='L77'>                Py_INCREF(<a href='../S/2252.html#L27' title='Defined at 27 in Modules/_ctypes/ctypes.h.'>value</a>);
<a name='L78'>                <a href='../S/2858.html#L294' title='Defined at 294 in PC/bdist_wininst/install.c.'>args</a> = <a href='../S/2252.html#L27' title='Defined at 27 in Modules/_ctypes/ctypes.h.'>value</a>;
<a name='L79'>            <font color='red'>}</font>
<a name='L80'>            <b>else</b>
<a name='L81'>                <a href='../S/2858.html#L294' title='Defined at 294 in PC/bdist_wininst/install.c.'>args</a> = <a href='../S/2823.html#L186' title='Defined at 186 in Objects/tupleobject.c.'>PyTuple_Pack</a>(1, <a href='../S/2252.html#L27' title='Defined at 27 in Modules/_ctypes/ctypes.h.'>value</a>);
<a name='L82'>            fixed_value = <a href='../S/2858.html#L294' title='Defined at 294 in PC/bdist_wininst/install.c.'>args</a> ?
<a name='L83'>                PyEval_CallObject(exception, <a href='../S/2858.html#L294' title='Defined at 294 in PC/bdist_wininst/install.c.'>args</a>) : <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>;
<a name='L84'>            <a href='../S/544.html#L767' title='Defined at 767 in Include/object.h.'>Py_XDECREF</a>(<a href='../S/2858.html#L294' title='Defined at 294 in PC/bdist_wininst/install.c.'>args</a>);
<a name='L85'>            <a href='../S/544.html#L767' title='Defined at 767 in Include/object.h.'>Py_XDECREF</a>(<a href='../S/2252.html#L27' title='Defined at 27 in Modules/_ctypes/ctypes.h.'>value</a>);
<a name='L86'>            <b>if</b> (fixed_value == <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>)
<a name='L87'>                <b>return</b>;
<a name='L88'>            <a href='../S/2252.html#L27' title='Defined at 27 in Modules/_ctypes/ctypes.h.'>value</a> = fixed_value;
<a name='L89'>        <font color='red'>}</font>
<a name='L90'>        <i><font color='green'>/* Avoid reference cycles through the context chain.</font></i>
<a name='L91'><i><font color='green'>           This is O(chain length) but context chains are</font></i>
<a name='L92'><i><font color='green'>           usually very short. Sensitive readers may try</font></i>
<a name='L93'><i><font color='green'>           to inline the call to PyException_GetContext. */</font></i>
<a name='L94'>        <b>if</b> (exc_value != <a href='../S/2252.html#L27' title='Defined at 27 in Modules/_ctypes/ctypes.h.'>value</a>) <font color='red'>{</font>
<a name='L95'>            <a href='../D/3593.html' title='Multiple defined in 2 places.'>PyObject</a> *o = exc_value, *context;
<a name='L96'>            <b>while</b> ((context = <a href='../S/2789.html#L346' title='Defined at 346 in Objects/exceptions.c.'>PyException_GetContext</a>(o))) <font color='red'>{</font>
<a name='L97'>                Py_DECREF(context);
<a name='L98'>                <b>if</b> (context == <a href='../S/2252.html#L27' title='Defined at 27 in Modules/_ctypes/ctypes.h.'>value</a>) <font color='red'>{</font>
<a name='L99'>                    <a href='../S/2789.html#L354' title='Defined at 354 in Objects/exceptions.c.'>PyException_SetContext</a>(o, <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>);
<a name='L100'>                    <b>break</b>;
<a name='L101'>                <font color='red'>}</font>
<a name='L102'>                o = context;
<a name='L103'>            <font color='red'>}</font>
<a name='L104'>            <a href='../S/2789.html#L354' title='Defined at 354 in Objects/exceptions.c.'>PyException_SetContext</a>(<a href='../S/2252.html#L27' title='Defined at 27 in Modules/_ctypes/ctypes.h.'>value</a>, exc_value);
<a name='L105'>        <font color='red'>}</font> <b>else</b> <font color='red'>{</font>
<a name='L106'>            Py_DECREF(exc_value);
<a name='L107'>        <font color='red'>}</font>
<a name='L108'>    <font color='red'>}</font>
<a name='L109'>    <b>if</b> (<a href='../S/2252.html#L27' title='Defined at 27 in Modules/_ctypes/ctypes.h.'>value</a> != <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a> &amp;&amp; PyExceptionInstance_Check(<a href='../S/2252.html#L27' title='Defined at 27 in Modules/_ctypes/ctypes.h.'>value</a>))
<a name='L110'>        tb = <a href='../S/2789.html#L318' title='Defined at 318 in Objects/exceptions.c.'>PyException_GetTraceback</a>(<a href='../S/2252.html#L27' title='Defined at 27 in Modules/_ctypes/ctypes.h.'>value</a>);
<a name='L111'>    <a href='../S/544.html#L766' title='Defined at 766 in Include/object.h.'>Py_XINCREF</a>(exception);
<a name='L112'>    <a href='../S/3106.html#L25' title='Defined at 25 in Python/errors.c.'>PyErr_Restore</a>(exception, <a href='../S/2252.html#L27' title='Defined at 27 in Modules/_ctypes/ctypes.h.'>value</a>, tb);
<a name='L113'><font color='red'>}</font>
<a name='L114'>
<a name='L115'><b>void</b>
<a name='L116'><a href='../R/2894.html' title='Multiple refered from 23 places.'>PyErr_SetNone</a>(PyObject *exception)
<a name='L117'><font color='red'>{</font>
<a name='L118'>    <a href='../S/3106.html#L53' title='Defined at 53 in Python/errors.c.'>PyErr_SetObject</a>(exception, (<a href='../D/3593.html' title='Multiple defined in 2 places.'>PyObject</a> *)<a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>);
<a name='L119'><font color='red'>}</font>
<a name='L120'>
<a name='L121'><b>void</b>
<a name='L122'><a href='../R/2896.html' title='Multiple refered from 1844 places.'>PyErr_SetString</a>(PyObject *exception, <b>const</b> <b>char</b> *string)
<a name='L123'><font color='red'>{</font>
<a name='L124'>    <a href='../D/3593.html' title='Multiple defined in 2 places.'>PyObject</a> *<a href='../S/2252.html#L27' title='Defined at 27 in Modules/_ctypes/ctypes.h.'>value</a> = <a href='../D/4012.html' title='Multiple defined in 3 places.'>PyUnicode_FromString</a>(<a href='../S/500.html#L5' title='Defined at 5 in Include/asdl.h.'>string</a>);
<a name='L125'>    <a href='../S/3106.html#L53' title='Defined at 53 in Python/errors.c.'>PyErr_SetObject</a>(exception, <a href='../S/2252.html#L27' title='Defined at 27 in Modules/_ctypes/ctypes.h.'>value</a>);
<a name='L126'>    <a href='../S/544.html#L767' title='Defined at 767 in Include/object.h.'>Py_XDECREF</a>(<a href='../S/2252.html#L27' title='Defined at 27 in Modules/_ctypes/ctypes.h.'>value</a>);
<a name='L127'><font color='red'>}</font>
<a name='L128'>
<a name='L129'>
<a name='L130'><a href='../D/3593.html' title='Multiple defined in 2 places.'>PyObject</a> *
<a name='L131'><a href='../R/2877.html' title='Multiple refered from 563 places.'>PyErr_Occurred</a>(<b>void</b>)
<a name='L132'><font color='red'>{</font>
<a name='L133'>    <i><font color='green'>/* If there is no thread state, PyThreadState_GET calls</font></i>
<a name='L134'><i><font color='green'>       Py_FatalError, which calls PyErr_Occurred.  To avoid the</font></i>
<a name='L135'><i><font color='green'>       resulting infinite loop, we inline PyThreadState_GET here and</font></i>
<a name='L136'><i><font color='green'>       treat no thread as no error. */</font></i>
<a name='L137'>    <a href='../D/3832.html' title='Multiple defined in 2 places.'>PyThreadState</a> *tstate =
<a name='L138'>        ((<a href='../D/3832.html' title='Multiple defined in 2 places.'>PyThreadState</a>*)<a href='../S/554.html#L173' title='Defined at 173 in Include/pyatomic.h.'>_Py_atomic_load_relaxed</a>(&amp;_PyThreadState_Current));
<a name='L139'>
<a name='L140'>    <b>return</b> tstate == <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a> ? <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a> : tstate-&gt;curexc_type;
<a name='L141'><font color='red'>}</font>
<a name='L142'>
<a name='L143'>
<a name='L144'><b>int</b>
<a name='L145'><a href='../R/2872.html' title='Multiple refered from 6 places.'>PyErr_GivenExceptionMatches</a>(PyObject *err, PyObject *exc)
<a name='L146'><font color='red'>{</font>
<a name='L147'>    <b>if</b> (err == <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a> || exc == <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>) <font color='red'>{</font>
<a name='L148'>        <i><font color='green'>/* maybe caused by "import exceptions" that failed early on */</font></i>
<a name='L149'>        <b>return</b> 0;
<a name='L150'>    <font color='red'>}</font>
<a name='L151'>    <b>if</b> (PyTuple_Check(exc)) <font color='red'>{</font>
<a name='L152'>        <a href='../D/4364.html' title='Multiple defined in 6 places.'>Py_ssize_t</a> i, n;
<a name='L153'>        n = <a href='../S/2823.html#L112' title='Defined at 112 in Objects/tupleobject.c.'>PyTuple_Size</a>(exc);
<a name='L154'>        <b>for</b> (i = 0; i &lt; n; i++) <font color='red'>{</font>
<a name='L155'>            <i><font color='green'>/* Test recursively */</font></i>
<a name='L156'>             <b>if</b> (<a href='../S/3106.html#L145' title='Defined at 145 in Python/errors.c.'>PyErr_GivenExceptionMatches</a>(
<a name='L157'>                 err, PyTuple_GET_ITEM(exc, i)))
<a name='L158'>             <font color='red'>{</font>
<a name='L159'>                 <b>return</b> 1;
<a name='L160'>             <font color='red'>}</font>
<a name='L161'>        <font color='red'>}</font>
<a name='L162'>        <b>return</b> 0;
<a name='L163'>    <font color='red'>}</font>
<a name='L164'>    <i><font color='green'>/* err might be an instance, so check its class. */</font></i>
<a name='L165'>    <b>if</b> (PyExceptionInstance_Check(err))
<a name='L166'>        err = PyExceptionInstance_Class(err);
<a name='L167'>
<a name='L168'>    <b>if</b> (PyExceptionClass_Check(err) &amp;&amp; PyExceptionClass_Check(exc)) <font color='red'>{</font>
<a name='L169'>        <b>int</b> res = 0;
<a name='L170'>        <a href='../D/3593.html' title='Multiple defined in 2 places.'>PyObject</a> *exception, *<a href='../S/2252.html#L27' title='Defined at 27 in Modules/_ctypes/ctypes.h.'>value</a>, *tb;
<a name='L171'>        <a href='../S/3106.html#L304' title='Defined at 304 in Python/errors.c.'>PyErr_Fetch</a>(&amp;exception, &amp;<a href='../S/2252.html#L27' title='Defined at 27 in Modules/_ctypes/ctypes.h.'>value</a>, &amp;tb);
<a name='L172'>        <i><font color='green'>/* PyObject_IsSubclass() can recurse and therefore is</font></i>
<a name='L173'><i><font color='green'>           not safe (see test_bad_getattr in test.pickletester). */</font></i>
<a name='L174'>        res = <a href='../S/2824.html#L1082' title='Defined at 1082 in Objects/typeobject.c.'>PyType_IsSubtype</a>((<a href='../D/3899.html' title='Multiple defined in 3 places.'>PyTypeObject</a> *)err, (<a href='../D/3899.html' title='Multiple defined in 3 places.'>PyTypeObject</a> *)exc);
<a name='L175'>        <i><font color='green'>/* This function must not fail, so print the error here */</font></i>
<a name='L176'>        <b>if</b> (res == -1) <font color='red'>{</font>
<a name='L177'>            <a href='../S/3106.html#L708' title='Defined at 708 in Python/errors.c.'>PyErr_WriteUnraisable</a>(err);
<a name='L178'>            res = 0;
<a name='L179'>        <font color='red'>}</font>
<a name='L180'>        <a href='../S/3106.html#L25' title='Defined at 25 in Python/errors.c.'>PyErr_Restore</a>(exception, <a href='../S/2252.html#L27' title='Defined at 27 in Modules/_ctypes/ctypes.h.'>value</a>, tb);
<a name='L181'>        <b>return</b> res;
<a name='L182'>    <font color='red'>}</font>
<a name='L183'>
<a name='L184'>    <b>return</b> err == exc;
<a name='L185'><font color='red'>}</font>
<a name='L186'>
<a name='L187'>
<a name='L188'><b>int</b>
<a name='L189'><a href='../R/2869.html' title='Multiple refered from 120 places.'>PyErr_ExceptionMatches</a>(PyObject *exc)
<a name='L190'><font color='red'>{</font>
<a name='L191'>    <b>return</b> <a href='../S/3106.html#L145' title='Defined at 145 in Python/errors.c.'>PyErr_GivenExceptionMatches</a>(<a href='../D/3154.html' title='Multiple defined in 2 places.'>PyErr_Occurred</a>(), exc);
<a name='L192'><font color='red'>}</font>
<a name='L193'>
<a name='L194'>
<a name='L195'><i><font color='green'>/* Used in many places to normalize a raised exception, including in</font></i>
<a name='L196'><i><font color='green'>   eval_code2(), do_raise(), and PyErr_Print()</font></i>
<a name='L197'><i><font color='green'></font></i>
<a name='L198'><i><font color='green'>   XXX: should PyErr_NormalizeException() also call</font></i>
<a name='L199'><i><font color='green'>            PyException_SetTraceback() with the resulting value and tb?</font></i>
<a name='L200'><i><font color='green'>*/</font></i>
<a name='L201'><b>void</b>
<a name='L202'><a href='../R/2876.html' title='Multiple refered from 9 places.'>PyErr_NormalizeException</a>(PyObject **exc, PyObject **val, PyObject **tb)
<a name='L203'><font color='red'>{</font>
<a name='L204'>    <a href='../D/3593.html' title='Multiple defined in 2 places.'>PyObject</a> *type = *exc;
<a name='L205'>    <a href='../D/3593.html' title='Multiple defined in 2 places.'>PyObject</a> *<a href='../S/2252.html#L27' title='Defined at 27 in Modules/_ctypes/ctypes.h.'>value</a> = *val;
<a name='L206'>    <a href='../D/3593.html' title='Multiple defined in 2 places.'>PyObject</a> *inclass = <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>;
<a name='L207'>    <a href='../D/3593.html' title='Multiple defined in 2 places.'>PyObject</a> *initial_tb = <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>;
<a name='L208'>    <a href='../D/3832.html' title='Multiple defined in 2 places.'>PyThreadState</a> *tstate = <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>;
<a name='L209'>
<a name='L210'>    <b>if</b> (type == <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>) <font color='red'>{</font>
<a name='L211'>        <i><font color='green'>/* There was no exception, so nothing to do. */</font></i>
<a name='L212'>        <b>return</b>;
<a name='L213'>    <font color='red'>}</font>
<a name='L214'>
<a name='L215'>    <i><font color='green'>/* If PyErr_SetNone() was used, the value will have been actually</font></i>
<a name='L216'><i><font color='green'>       set to NULL.</font></i>
<a name='L217'><i><font color='green'>    */</font></i>
<a name='L218'>    <b>if</b> (!<a href='../S/2252.html#L27' title='Defined at 27 in Modules/_ctypes/ctypes.h.'>value</a>) <font color='red'>{</font>
<a name='L219'>        <a href='../S/2252.html#L27' title='Defined at 27 in Modules/_ctypes/ctypes.h.'>value</a> = Py_None;
<a name='L220'>        Py_INCREF(<a href='../S/2252.html#L27' title='Defined at 27 in Modules/_ctypes/ctypes.h.'>value</a>);
<a name='L221'>    <font color='red'>}</font>
<a name='L222'>
<a name='L223'>    <b>if</b> (PyExceptionInstance_Check(<a href='../S/2252.html#L27' title='Defined at 27 in Modules/_ctypes/ctypes.h.'>value</a>))
<a name='L224'>        inclass = PyExceptionInstance_Class(<a href='../S/2252.html#L27' title='Defined at 27 in Modules/_ctypes/ctypes.h.'>value</a>);
<a name='L225'>
<a name='L226'>    <i><font color='green'>/* Normalize the exception so that if the type is a class, the</font></i>
<a name='L227'><i><font color='green'>       value will be an instance.</font></i>
<a name='L228'><i><font color='green'>    */</font></i>
<a name='L229'>    <b>if</b> (PyExceptionClass_Check(type)) <font color='red'>{</font>
<a name='L230'>        <i><font color='green'>/* if the value was not an instance, or is not an instance</font></i>
<a name='L231'><i><font color='green'>           whose class is (or is derived from) type, then use the</font></i>
<a name='L232'><i><font color='green'>           value as an argument to instantiation of the type</font></i>
<a name='L233'><i><font color='green'>           class.</font></i>
<a name='L234'><i><font color='green'>        */</font></i>
<a name='L235'>        <b>if</b> (!inclass || !<a href='../S/2775.html#L2608' title='Defined at 2608 in Objects/abstract.c.'>PyObject_IsSubclass</a>(inclass, type)) <font color='red'>{</font>
<a name='L236'>            <a href='../D/3593.html' title='Multiple defined in 2 places.'>PyObject</a> *<a href='../S/2858.html#L294' title='Defined at 294 in PC/bdist_wininst/install.c.'>args</a>, *res;
<a name='L237'>
<a name='L238'>            <b>if</b> (<a href='../S/2252.html#L27' title='Defined at 27 in Modules/_ctypes/ctypes.h.'>value</a> == Py_None)
<a name='L239'>                <a href='../S/2858.html#L294' title='Defined at 294 in PC/bdist_wininst/install.c.'>args</a> = <a href='../S/2823.html#L49' title='Defined at 49 in Objects/tupleobject.c.'>PyTuple_New</a>(0);
<a name='L240'>            <b>else</b> <b>if</b> (PyTuple_Check(<a href='../S/2252.html#L27' title='Defined at 27 in Modules/_ctypes/ctypes.h.'>value</a>)) <font color='red'>{</font>
<a name='L241'>                Py_INCREF(<a href='../S/2252.html#L27' title='Defined at 27 in Modules/_ctypes/ctypes.h.'>value</a>);
<a name='L242'>                <a href='../S/2858.html#L294' title='Defined at 294 in PC/bdist_wininst/install.c.'>args</a> = <a href='../S/2252.html#L27' title='Defined at 27 in Modules/_ctypes/ctypes.h.'>value</a>;
<a name='L243'>            <font color='red'>}</font>
<a name='L244'>            <b>else</b>
<a name='L245'>                <a href='../S/2858.html#L294' title='Defined at 294 in PC/bdist_wininst/install.c.'>args</a> = <a href='../S/2823.html#L186' title='Defined at 186 in Objects/tupleobject.c.'>PyTuple_Pack</a>(1, <a href='../S/2252.html#L27' title='Defined at 27 in Modules/_ctypes/ctypes.h.'>value</a>);
<a name='L246'>
<a name='L247'>            <b>if</b> (<a href='../S/2858.html#L294' title='Defined at 294 in PC/bdist_wininst/install.c.'>args</a> == <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>)
<a name='L248'>                <b>goto</b> finally;
<a name='L249'>            res = PyEval_CallObject(type, <a href='../S/2858.html#L294' title='Defined at 294 in PC/bdist_wininst/install.c.'>args</a>);
<a name='L250'>            Py_DECREF(<a href='../S/2858.html#L294' title='Defined at 294 in PC/bdist_wininst/install.c.'>args</a>);
<a name='L251'>            <b>if</b> (res == <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>)
<a name='L252'>                <b>goto</b> finally;
<a name='L253'>            Py_DECREF(<a href='../S/2252.html#L27' title='Defined at 27 in Modules/_ctypes/ctypes.h.'>value</a>);
<a name='L254'>            <a href='../S/2252.html#L27' title='Defined at 27 in Modules/_ctypes/ctypes.h.'>value</a> = res;
<a name='L255'>        <font color='red'>}</font>
<a name='L256'>        <i><font color='green'>/* if the class of the instance doesn't exactly match the</font></i>
<a name='L257'><i><font color='green'>           class of the type, believe the instance</font></i>
<a name='L258'><i><font color='green'>        */</font></i>
<a name='L259'>        <b>else</b> <b>if</b> (inclass != type) <font color='red'>{</font>
<a name='L260'>            Py_DECREF(type);
<a name='L261'>            type = inclass;
<a name='L262'>            Py_INCREF(type);
<a name='L263'>        <font color='red'>}</font>
<a name='L264'>    <font color='red'>}</font>
<a name='L265'>    *exc = type;
<a name='L266'>    *val = <a href='../S/2252.html#L27' title='Defined at 27 in Modules/_ctypes/ctypes.h.'>value</a>;
<a name='L267'>    <b>return</b>;
<a name='L268'>finally:
<a name='L269'>    Py_DECREF(type);
<a name='L270'>    Py_DECREF(<a href='../S/2252.html#L27' title='Defined at 27 in Modules/_ctypes/ctypes.h.'>value</a>);
<a name='L271'>    <i><font color='green'>/* If the new exception doesn't set a traceback and the old</font></i>
<a name='L272'><i><font color='green'>       exception had a traceback, use the old traceback for the</font></i>
<a name='L273'><i><font color='green'>       new exception.  It's better than nothing.</font></i>
<a name='L274'><i><font color='green'>    */</font></i>
<a name='L275'>    initial_tb = *tb;
<a name='L276'>    <a href='../S/3106.html#L304' title='Defined at 304 in Python/errors.c.'>PyErr_Fetch</a>(exc, val, tb);
<a name='L277'>    <b>if</b> (initial_tb != <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>) <font color='red'>{</font>
<a name='L278'>        <b>if</b> (*tb == <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>)
<a name='L279'>            *tb = initial_tb;
<a name='L280'>        <b>else</b>
<a name='L281'>            Py_DECREF(initial_tb);
<a name='L282'>    <font color='red'>}</font>
<a name='L283'>    <i><font color='green'>/* normalize recursively */</font></i>
<a name='L284'>    tstate = <a href='../S/509.html#L90' title='Defined at 90 in Include/ceval.h.'>PyThreadState_GET</a>();
<a name='L285'>    <b>if</b> (++tstate-&gt;recursion_depth &gt; <a href='../S/3091.html#L683' title='Defined at 683 in Python/ceval.c.'>Py_GetRecursionLimit</a>()) <font color='red'>{</font>
<a name='L286'>        --tstate-&gt;recursion_depth;
<a name='L287'>        <i><font color='green'>/* throw away the old exception... */</font></i>
<a name='L288'>        Py_DECREF(*exc);
<a name='L289'>        Py_DECREF(*val);
<a name='L290'>        <i><font color='green'>/* ... and use the recursion error instead */</font></i>
<a name='L291'>        *exc = PyExc_RuntimeError;
<a name='L292'>        *val = PyExc_RecursionErrorInst;
<a name='L293'>        Py_INCREF(*exc);
<a name='L294'>        Py_INCREF(*val);
<a name='L295'>        <i><font color='green'>/* just keeping the old traceback */</font></i>
<a name='L296'>        <b>return</b>;
<a name='L297'>    <font color='red'>}</font>
<a name='L298'>    <a href='../S/3106.html#L202' title='Defined at 202 in Python/errors.c.'>PyErr_NormalizeException</a>(exc, val, tb);
<a name='L299'>    --tstate-&gt;recursion_depth;
<a name='L300'><font color='red'>}</font>
<a name='L301'>
<a name='L302'>
<a name='L303'><b>void</b>
<a name='L304'><a href='../R/2870.html' title='Multiple refered from 39 places.'>PyErr_Fetch</a>(PyObject **p_type, PyObject **p_value, PyObject **p_traceback)
<a name='L305'><font color='red'>{</font>
<a name='L306'>    <a href='../D/3832.html' title='Multiple defined in 2 places.'>PyThreadState</a> *tstate = <a href='../S/509.html#L90' title='Defined at 90 in Include/ceval.h.'>PyThreadState_GET</a>();
<a name='L307'>
<a name='L308'>    *p_type = tstate-&gt;curexc_type;
<a name='L309'>    *p_value = tstate-&gt;curexc_value;
<a name='L310'>    *p_traceback = tstate-&gt;curexc_traceback;
<a name='L311'>
<a name='L312'>    tstate-&gt;curexc_type = <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>;
<a name='L313'>    tstate-&gt;curexc_value = <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>;
<a name='L314'>    tstate-&gt;curexc_traceback = <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>;
<a name='L315'><font color='red'>}</font>
<a name='L316'>
<a name='L317'><b>void</b>
<a name='L318'><a href='../R/2867.html' title='Multiple refered from 352 places.'>PyErr_Clear</a>(<b>void</b>)
<a name='L319'><font color='red'>{</font>
<a name='L320'>    <a href='../S/3106.html#L25' title='Defined at 25 in Python/errors.c.'>PyErr_Restore</a>(<a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>, <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>, <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>);
<a name='L321'><font color='red'>}</font>
<a name='L322'>
<a name='L323'><i><font color='green'>/* Convenience functions to set a type error exception and return 0 */</font></i>
<a name='L324'>
<a name='L325'><b>int</b>
<a name='L326'><a href='../R/2864.html' title='Multiple refered from 41 places.'>PyErr_BadArgument</a>(<b>void</b>)
<a name='L327'><font color='red'>{</font>
<a name='L328'>    <a href='../S/3106.html#L122' title='Defined at 122 in Python/errors.c.'>PyErr_SetString</a>(PyExc_TypeError,
<a name='L329'>                    "bad argument type for built-in operation");
<a name='L330'>    <b>return</b> 0;
<a name='L331'><font color='red'>}</font>
<a name='L332'>
<a name='L333'><a href='../D/3593.html' title='Multiple defined in 2 places.'>PyObject</a> *
<a name='L334'><a href='../R/2875.html' title='Multiple refered from 271 places.'>PyErr_NoMemory</a>(<b>void</b>)
<a name='L335'><font color='red'>{</font>
<a name='L336'>    <a href='../S/3106.html#L116' title='Defined at 116 in Python/errors.c.'>PyErr_SetNone</a>(PyExc_MemoryError);
<a name='L337'>    <b>return</b> <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>;
<a name='L338'><font color='red'>}</font>
<a name='L339'>
<a name='L340'><a href='../D/3593.html' title='Multiple defined in 2 places.'>PyObject</a> *
<a name='L341'><a href='../R/2887.html' title='Multiple refered from 5 places.'>PyErr_SetFromErrnoWithFilenameObject</a>(PyObject *exc, PyObject *filenameObject)
<a name='L342'><font color='red'>{</font>
<a name='L343'>    <a href='../D/3593.html' title='Multiple defined in 2 places.'>PyObject</a> *message;
<a name='L344'>    <a href='../D/3593.html' title='Multiple defined in 2 places.'>PyObject</a> *v;
<a name='L345'>    <b>int</b> i = errno;
<a name='L346'><font color='darkred'>#ifndef</font> <a href='../S/2906.html#L81' title='Defined at 81 in PC/pyconfig.h.'>MS_WINDOWS</a>
<a name='L347'>    <b>char</b> *s;
<a name='L348'><font color='darkred'>#else</font>
<a name='L349'>    WCHAR *s_buf = <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>;
<a name='L350'><font color='darkred'>#endif</font> <i><font color='green'>/* Unix/Windows */</font></i>
<a name='L351'>
<a name='L352'><font color='darkred'>#ifdef</font> EINTR
<a name='L353'>    <b>if</b> (i == EINTR &amp;&amp; <a href='../D/3143.html' title='Multiple defined in 2 places.'>PyErr_CheckSignals</a>())
<a name='L354'>        <b>return</b> <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>;
<a name='L355'><font color='darkred'>#endif</font>
<a name='L356'>
<a name='L357'><font color='darkred'>#ifndef</font> <a href='../S/2906.html#L81' title='Defined at 81 in PC/pyconfig.h.'>MS_WINDOWS</a>
<a name='L358'>    <b>if</b> (i == 0)
<a name='L359'>        s = "Error"; <i><font color='green'>/* Sometimes errno didn't get set */</font></i>
<a name='L360'>    <b>else</b>
<a name='L361'>        s = strerror(i);
<a name='L362'>    message = <a href='../D/3987.html' title='Multiple defined in 3 places.'>PyUnicode_DecodeUTF8</a>(s, strlen(s), "ignore");
<a name='L363'><font color='darkred'>#else</font>
<a name='L364'>    <b>if</b> (i == 0)
<a name='L365'>        message = <a href='../D/4012.html' title='Multiple defined in 3 places.'>PyUnicode_FromString</a>("Error"); <i><font color='green'>/* Sometimes errno didn't get set */</font></i>
<a name='L366'>    <b>else</b>
<a name='L367'>    <font color='red'>{</font>
<a name='L368'>        <i><font color='green'>/* Note that the Win32 errors do not lineup with the</font></i>
<a name='L369'><i><font color='green'>           errno error.  So if the error is in the MSVC error</font></i>
<a name='L370'><i><font color='green'>           table, we use it, otherwise we assume it really _is_</font></i>
<a name='L371'><i><font color='green'>           a Win32 error code</font></i>
<a name='L372'><i><font color='green'>        */</font></i>
<a name='L373'>        <b>if</b> (i &gt; 0 &amp;&amp; i &lt; _sys_nerr) <font color='red'>{</font>
<a name='L374'>            message = <a href='../D/4012.html' title='Multiple defined in 3 places.'>PyUnicode_FromString</a>(_sys_errlist[i]);
<a name='L375'>        <font color='red'>}</font>
<a name='L376'>        <b>else</b> <font color='red'>{</font>
<a name='L377'>            <b>int</b> len = FormatMessageW(
<a name='L378'>                FORMAT_MESSAGE_ALLOCATE_BUFFER |
<a name='L379'>                FORMAT_MESSAGE_FROM_SYSTEM |
<a name='L380'>                FORMAT_MESSAGE_IGNORE_INSERTS,
<a name='L381'>                <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>,                   <i><font color='green'>/* no message source */</font></i>
<a name='L382'>                i,
<a name='L383'>                MAKELANGID(LANG_NEUTRAL,
<a name='L384'>                           SUBLANG_DEFAULT),
<a name='L385'>                           <i><font color='green'>/* Default language */</font></i>
<a name='L386'>                (LPWSTR) &amp;s_buf,
<a name='L387'>                0,                      <i><font color='green'>/* size not used */</font></i>
<a name='L388'>                <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>);                  <i><font color='green'>/* no args */</font></i>
<a name='L389'>            <b>if</b> (len==0) <font color='red'>{</font>
<a name='L390'>                <i><font color='green'>/* Only ever seen this in out-of-mem</font></i>
<a name='L391'><i><font color='green'>                   situations */</font></i>
<a name='L392'>                s_buf = <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>;
<a name='L393'>                message = <a href='../D/4008.html' title='Multiple defined in 3 places.'>PyUnicode_FromFormat</a>("Windows Error 0x%X", i);
<a name='L394'>            <font color='red'>}</font> <b>else</b> <font color='red'>{</font>
<a name='L395'>                <i><font color='green'>/* remove trailing cr/lf and dots */</font></i>
<a name='L396'>                <b>while</b> (len &gt; 0 &amp;&amp; (s_buf[len-1] &lt;= L' ' || s_buf[len-1] == L'.'))
<a name='L397'>                    s_buf[--len] = L'\0';
<a name='L398'>                message = <a href='../D/4014.html' title='Multiple defined in 3 places.'>PyUnicode_FromUnicode</a>(s_buf, len);
<a name='L399'>            <font color='red'>}</font>
<a name='L400'>        <font color='red'>}</font>
<a name='L401'>    <font color='red'>}</font>
<a name='L402'><font color='darkred'>#endif</font> <i><font color='green'>/* Unix/Windows */</font></i>
<a name='L403'>
<a name='L404'>    <b>if</b> (message == <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>)
<a name='L405'>    <font color='red'>{</font>
<a name='L406'><font color='darkred'>#ifdef</font> <a href='../S/2906.html#L81' title='Defined at 81 in PC/pyconfig.h.'>MS_WINDOWS</a>
<a name='L407'>        LocalFree(s_buf);
<a name='L408'><font color='darkred'>#endif</font>
<a name='L409'>        <b>return</b> <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>;
<a name='L410'>    <font color='red'>}</font>
<a name='L411'>
<a name='L412'>    <b>if</b> (filenameObject != <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>)
<a name='L413'>        v = <a href='../D/4089.html' title='Multiple defined in 2 places.'>Py_BuildValue</a>("(iOO)", i, message, filenameObject);
<a name='L414'>    <b>else</b>
<a name='L415'>        v = <a href='../D/4089.html' title='Multiple defined in 2 places.'>Py_BuildValue</a>("(iO)", i, message);
<a name='L416'>    Py_DECREF(message);
<a name='L417'>
<a name='L418'>    <b>if</b> (v != <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>) <font color='red'>{</font>
<a name='L419'>        <a href='../S/3106.html#L53' title='Defined at 53 in Python/errors.c.'>PyErr_SetObject</a>(exc, v);
<a name='L420'>        Py_DECREF(v);
<a name='L421'>    <font color='red'>}</font>
<a name='L422'><font color='darkred'>#ifdef</font> <a href='../S/2906.html#L81' title='Defined at 81 in PC/pyconfig.h.'>MS_WINDOWS</a>
<a name='L423'>    LocalFree(s_buf);
<a name='L424'><font color='darkred'>#endif</font>
<a name='L425'>    <b>return</b> <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>;
<a name='L426'><font color='red'>}</font>
<a name='L427'>
<a name='L428'>
<a name='L429'><a href='../D/3593.html' title='Multiple defined in 2 places.'>PyObject</a> *
<a name='L430'><a href='../R/2886.html' title='Multiple refered from 13 places.'>PyErr_SetFromErrnoWithFilename</a>(PyObject *exc, <b>const</b> <b>char</b> *filename)
<a name='L431'><font color='red'>{</font>
<a name='L432'>    <a href='../D/3593.html' title='Multiple defined in 2 places.'>PyObject</a> *name = filename ? <a href='../D/3975.html' title='Multiple defined in 3 places.'>PyUnicode_DecodeFSDefault</a>(filename) : <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>;
<a name='L433'>    <a href='../D/3593.html' title='Multiple defined in 2 places.'>PyObject</a> *<a href='../S/2250.html#L579' title='Defined at 579 in Modules/_ctypes/callproc.c.'>result</a> = <a href='../S/3106.html#L341' title='Defined at 341 in Python/errors.c.'>PyErr_SetFromErrnoWithFilenameObject</a>(exc, name);
<a name='L434'>    <a href='../S/544.html#L767' title='Defined at 767 in Include/object.h.'>Py_XDECREF</a>(name);
<a name='L435'>    <b>return</b> <a href='../S/2250.html#L579' title='Defined at 579 in Modules/_ctypes/callproc.c.'>result</a>;
<a name='L436'><font color='red'>}</font>
<a name='L437'>
<a name='L438'><font color='darkred'>#ifdef</font> <a href='../S/2906.html#L81' title='Defined at 81 in PC/pyconfig.h.'>MS_WINDOWS</a>
<a name='L439'><a href='../D/3593.html' title='Multiple defined in 2 places.'>PyObject</a> *
<a name='L440'><a href='../S/2572.html#L381' title='Refered from 381 in Modules/_io/fileio.c.'>PyErr_SetFromErrnoWithUnicodeFilename</a>(PyObject *exc, <b>const</b> Py_UNICODE *filename)
<a name='L441'><font color='red'>{</font>
<a name='L442'>    <a href='../D/3593.html' title='Multiple defined in 2 places.'>PyObject</a> *name = filename ?
<a name='L443'>                     <a href='../D/4014.html' title='Multiple defined in 3 places.'>PyUnicode_FromUnicode</a>(filename, wcslen(filename)) :
<a name='L444'>             <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>;
<a name='L445'>    <a href='../D/3593.html' title='Multiple defined in 2 places.'>PyObject</a> *<a href='../S/2250.html#L579' title='Defined at 579 in Modules/_ctypes/callproc.c.'>result</a> = <a href='../S/3106.html#L341' title='Defined at 341 in Python/errors.c.'>PyErr_SetFromErrnoWithFilenameObject</a>(exc, name);
<a name='L446'>    <a href='../S/544.html#L767' title='Defined at 767 in Include/object.h.'>Py_XDECREF</a>(name);
<a name='L447'>    <b>return</b> <a href='../S/2250.html#L579' title='Defined at 579 in Modules/_ctypes/callproc.c.'>result</a>;
<a name='L448'><font color='red'>}</font>
<a name='L449'><font color='darkred'>#endif</font> <i><font color='green'>/* MS_WINDOWS */</font></i>
<a name='L450'>
<a name='L451'><a href='../D/3593.html' title='Multiple defined in 2 places.'>PyObject</a> *
<a name='L452'><a href='../R/2885.html' title='Multiple refered from 121 places.'>PyErr_SetFromErrno</a>(PyObject *exc)
<a name='L453'><font color='red'>{</font>
<a name='L454'>    <b>return</b> <a href='../S/3106.html#L341' title='Defined at 341 in Python/errors.c.'>PyErr_SetFromErrnoWithFilenameObject</a>(exc, <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>);
<a name='L455'><font color='red'>}</font>
<a name='L456'>
<a name='L457'><font color='darkred'>#ifdef</font> <a href='../S/2906.html#L81' title='Defined at 81 in PC/pyconfig.h.'>MS_WINDOWS</a>
<a name='L458'><i><font color='green'>/* Windows specific error code handling */</font></i>
<a name='L459'><a href='../D/3593.html' title='Multiple defined in 2 places.'>PyObject</a> *<a href='../R/2884.html' title='Multiple refered from 4 places.'>PyErr_SetExcFromWindowsErrWithFilenameObject</a>(
<a name='L460'>    PyObject *exc,
<a name='L461'>    <b>int</b> ierr,
<a name='L462'>    PyObject *filenameObject)
<a name='L463'><font color='red'>{</font>
<a name='L464'>    <b>int</b> len;
<a name='L465'>    WCHAR *s_buf = <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>; <i><font color='green'>/* Free via LocalFree */</font></i>
<a name='L466'>    <a href='../D/3593.html' title='Multiple defined in 2 places.'>PyObject</a> *message;
<a name='L467'>    <a href='../D/3593.html' title='Multiple defined in 2 places.'>PyObject</a> *v;
<a name='L468'>    DWORD err = (DWORD)ierr;
<a name='L469'>    <b>if</b> (err==0) err = GetLastError();
<a name='L470'>    len = FormatMessageW(
<a name='L471'>        <i><font color='green'>/* Error API error */</font></i>
<a name='L472'>        FORMAT_MESSAGE_ALLOCATE_BUFFER |
<a name='L473'>        FORMAT_MESSAGE_FROM_SYSTEM |
<a name='L474'>        FORMAT_MESSAGE_IGNORE_INSERTS,
<a name='L475'>        <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>,           <i><font color='green'>/* no message source */</font></i>
<a name='L476'>        err,
<a name='L477'>        MAKELANGID(LANG_NEUTRAL,
<a name='L478'>        SUBLANG_DEFAULT), <i><font color='green'>/* Default language */</font></i>
<a name='L479'>        (LPWSTR) &amp;s_buf,
<a name='L480'>        0,              <i><font color='green'>/* size not used */</font></i>
<a name='L481'>        <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>);          <i><font color='green'>/* no args */</font></i>
<a name='L482'>    <b>if</b> (len==0) <font color='red'>{</font>
<a name='L483'>        <i><font color='green'>/* Only seen this in out of mem situations */</font></i>
<a name='L484'>        message = <a href='../D/4008.html' title='Multiple defined in 3 places.'>PyUnicode_FromFormat</a>("Windows Error 0x%X", err);
<a name='L485'>        s_buf = <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>;
<a name='L486'>    <font color='red'>}</font> <b>else</b> <font color='red'>{</font>
<a name='L487'>        <i><font color='green'>/* remove trailing cr/lf and dots */</font></i>
<a name='L488'>        <b>while</b> (len &gt; 0 &amp;&amp; (s_buf[len-1] &lt;= L' ' || s_buf[len-1] == L'.'))
<a name='L489'>            s_buf[--len] = L'\0';
<a name='L490'>        message = <a href='../D/4014.html' title='Multiple defined in 3 places.'>PyUnicode_FromUnicode</a>(s_buf, len);
<a name='L491'>    <font color='red'>}</font>
<a name='L492'>
<a name='L493'>    <b>if</b> (message == <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>)
<a name='L494'>    <font color='red'>{</font>
<a name='L495'>        LocalFree(s_buf);
<a name='L496'>        <b>return</b> <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>;
<a name='L497'>    <font color='red'>}</font>
<a name='L498'>
<a name='L499'>    <b>if</b> (filenameObject != <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>)
<a name='L500'>        v = <a href='../D/4089.html' title='Multiple defined in 2 places.'>Py_BuildValue</a>("(iOO)", err, message, filenameObject);
<a name='L501'>    <b>else</b>
<a name='L502'>        v = <a href='../D/4089.html' title='Multiple defined in 2 places.'>Py_BuildValue</a>("(iO)", err, message);
<a name='L503'>    Py_DECREF(message);
<a name='L504'>
<a name='L505'>    <b>if</b> (v != <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>) <font color='red'>{</font>
<a name='L506'>        <a href='../S/3106.html#L53' title='Defined at 53 in Python/errors.c.'>PyErr_SetObject</a>(exc, v);
<a name='L507'>        Py_DECREF(v);
<a name='L508'>    <font color='red'>}</font>
<a name='L509'>    LocalFree(s_buf);
<a name='L510'>    <b>return</b> <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>;
<a name='L511'><font color='red'>}</font>
<a name='L512'>
<a name='L513'><a href='../D/3593.html' title='Multiple defined in 2 places.'>PyObject</a> *<a href='../R/2883.html' title='Multiple refered from 2 places.'>PyErr_SetExcFromWindowsErrWithFilename</a>(
<a name='L514'>    PyObject *exc,
<a name='L515'>    <b>int</b> ierr,
<a name='L516'>    <b>const</b> <b>char</b> *filename)
<a name='L517'><font color='red'>{</font>
<a name='L518'>    <a href='../D/3593.html' title='Multiple defined in 2 places.'>PyObject</a> *name = filename ? <a href='../D/3975.html' title='Multiple defined in 3 places.'>PyUnicode_DecodeFSDefault</a>(filename) : <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>;
<a name='L519'>    <a href='../D/3593.html' title='Multiple defined in 2 places.'>PyObject</a> *ret = <a href='../S/3106.html#L459' title='Defined at 459 in Python/errors.c.'>PyErr_SetExcFromWindowsErrWithFilenameObject</a>(exc,
<a name='L520'>                                                                 ierr,
<a name='L521'>                                                                 name);
<a name='L522'>    <a href='../S/544.html#L767' title='Defined at 767 in Include/object.h.'>Py_XDECREF</a>(name);
<a name='L523'>    <b>return</b> ret;
<a name='L524'><font color='red'>}</font>
<a name='L525'>
<a name='L526'><a href='../D/3593.html' title='Multiple defined in 2 places.'>PyObject</a> *PyErr_SetExcFromWindowsErrWithUnicodeFilename(
<a name='L527'>    PyObject *exc,
<a name='L528'>    <b>int</b> ierr,
<a name='L529'>    <b>const</b> Py_UNICODE *filename)
<a name='L530'><font color='red'>{</font>
<a name='L531'>    <a href='../D/3593.html' title='Multiple defined in 2 places.'>PyObject</a> *name = filename ?
<a name='L532'>                     <a href='../D/4014.html' title='Multiple defined in 3 places.'>PyUnicode_FromUnicode</a>(filename, wcslen(filename)) :
<a name='L533'>             <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>;
<a name='L534'>    <a href='../D/3593.html' title='Multiple defined in 2 places.'>PyObject</a> *ret = <a href='../S/3106.html#L459' title='Defined at 459 in Python/errors.c.'>PyErr_SetExcFromWindowsErrWithFilenameObject</a>(exc,
<a name='L535'>                                                                 ierr,
<a name='L536'>                                                                 name);
<a name='L537'>    <a href='../S/544.html#L767' title='Defined at 767 in Include/object.h.'>Py_XDECREF</a>(name);
<a name='L538'>    <b>return</b> ret;
<a name='L539'><font color='red'>}</font>
<a name='L540'>
<a name='L541'><a href='../D/3593.html' title='Multiple defined in 2 places.'>PyObject</a> *<a href='../R/2882.html' title='Multiple refered from 5 places.'>PyErr_SetExcFromWindowsErr</a>(PyObject *exc, <b>int</b> ierr)
<a name='L542'><font color='red'>{</font>
<a name='L543'>    <b>return</b> <a href='../S/3106.html#L513' title='Defined at 513 in Python/errors.c.'>PyErr_SetExcFromWindowsErrWithFilename</a>(exc, ierr, <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>);
<a name='L544'><font color='red'>}</font>
<a name='L545'>
<a name='L546'><a href='../D/3593.html' title='Multiple defined in 2 places.'>PyObject</a> *<a href='../R/2889.html' title='Multiple refered from 39 places.'>PyErr_SetFromWindowsErr</a>(<b>int</b> ierr)
<a name='L547'><font color='red'>{</font>
<a name='L548'>    <b>return</b> <a href='../S/3106.html#L513' title='Defined at 513 in Python/errors.c.'>PyErr_SetExcFromWindowsErrWithFilename</a>(PyExc_WindowsError,
<a name='L549'>                                                  ierr, <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>);
<a name='L550'><font color='red'>}</font>
<a name='L551'><a href='../D/3593.html' title='Multiple defined in 2 places.'>PyObject</a> *<a href='../R/2890.html' title='Multiple refered from 4 places.'>PyErr_SetFromWindowsErrWithFilename</a>(
<a name='L552'>    <b>int</b> ierr,
<a name='L553'>    <b>const</b> <b>char</b> *filename)
<a name='L554'><font color='red'>{</font>
<a name='L555'>    <a href='../D/3593.html' title='Multiple defined in 2 places.'>PyObject</a> *name = filename ? <a href='../D/3975.html' title='Multiple defined in 3 places.'>PyUnicode_DecodeFSDefault</a>(filename) : <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>;
<a name='L556'>    <a href='../D/3593.html' title='Multiple defined in 2 places.'>PyObject</a> *<a href='../S/2250.html#L579' title='Defined at 579 in Modules/_ctypes/callproc.c.'>result</a> = <a href='../S/3106.html#L459' title='Defined at 459 in Python/errors.c.'>PyErr_SetExcFromWindowsErrWithFilenameObject</a>(
<a name='L557'>                                                  PyExc_WindowsError,
<a name='L558'>                                                  ierr, name);
<a name='L559'>    <a href='../S/544.html#L767' title='Defined at 767 in Include/object.h.'>Py_XDECREF</a>(name);
<a name='L560'>    <b>return</b> <a href='../S/2250.html#L579' title='Defined at 579 in Modules/_ctypes/callproc.c.'>result</a>;
<a name='L561'><font color='red'>}</font>
<a name='L562'>
<a name='L563'><a href='../D/3593.html' title='Multiple defined in 2 places.'>PyObject</a> *<a href='../S/2696.html#L695' title='Refered from 695 in Modules/posixmodule.c.'>PyErr_SetFromWindowsErrWithUnicodeFilename</a>(
<a name='L564'>    <b>int</b> ierr,
<a name='L565'>    <b>const</b> Py_UNICODE *filename)
<a name='L566'><font color='red'>{</font>
<a name='L567'>    <a href='../D/3593.html' title='Multiple defined in 2 places.'>PyObject</a> *name = filename ?
<a name='L568'>                     <a href='../D/4014.html' title='Multiple defined in 3 places.'>PyUnicode_FromUnicode</a>(filename, wcslen(filename)) :
<a name='L569'>             <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>;
<a name='L570'>    <a href='../D/3593.html' title='Multiple defined in 2 places.'>PyObject</a> *<a href='../S/2250.html#L579' title='Defined at 579 in Modules/_ctypes/callproc.c.'>result</a> = <a href='../S/3106.html#L459' title='Defined at 459 in Python/errors.c.'>PyErr_SetExcFromWindowsErrWithFilenameObject</a>(
<a name='L571'>                                                  PyExc_WindowsError,
<a name='L572'>                                                  ierr, name);
<a name='L573'>    <a href='../S/544.html#L767' title='Defined at 767 in Include/object.h.'>Py_XDECREF</a>(name);
<a name='L574'>    <b>return</b> <a href='../S/2250.html#L579' title='Defined at 579 in Modules/_ctypes/callproc.c.'>result</a>;
<a name='L575'><font color='red'>}</font>
<a name='L576'><font color='darkred'>#endif</font> <i><font color='green'>/* MS_WINDOWS */</font></i>
<a name='L577'>
<a name='L578'><b>void</b>
<a name='L579'><a href='../S/3106.html#L595' title='Refered from 595 in Python/errors.c.'>_PyErr_BadInternalCall</a>(<b>const</b> <b>char</b> *filename, <b>int</b> lineno)
<a name='L580'><font color='red'>{</font>
<a name='L581'>    <a href='../S/3106.html#L600' title='Defined at 600 in Python/errors.c.'>PyErr_Format</a>(PyExc_SystemError,
<a name='L582'>                 "%s:%d: bad argument to internal function",
<a name='L583'>                 filename, lineno);
<a name='L584'><font color='red'>}</font>
<a name='L585'>
<a name='L586'><i><font color='green'>/* Remove the preprocessor macro for PyErr_BadInternalCall() so that we can</font></i>
<a name='L587'><i><font color='green'>   export the entry point for existing object code: */</font></i>
<a name='L588'><font color='darkred'>#undef</font> <a href='../R/2865.html' title='Multiple refered from 120 places.'>PyErr_BadInternalCall</a>
<a name='L589'><b>void</b>
<a name='L590'><a href='../R/2865.html' title='Multiple refered from 120 places.'>PyErr_BadInternalCall</a>(<b>void</b>)
<a name='L591'><font color='red'>{</font>
<a name='L592'>    <a href='../S/3106.html#L600' title='Defined at 600 in Python/errors.c.'>PyErr_Format</a>(PyExc_SystemError,
<a name='L593'>                 "bad argument to internal function");
<a name='L594'><font color='red'>}</font>
<a name='L595'><font color='darkred'>#define</font> <a href='../R/2865.html' title='Multiple refered from 120 places.'>PyErr_BadInternalCall</a>() <a href='../S/3106.html#L579' title='Defined at 579 in Python/errors.c.'>_PyErr_BadInternalCall</a>(__FILE__, __LINE__)
<a name='L596'>
<a name='L597'>
<a name='L598'>
<a name='L599'><a href='../D/3593.html' title='Multiple defined in 2 places.'>PyObject</a> *
<a name='L600'><a href='../R/2871.html' title='Multiple refered from 690 places.'>PyErr_Format</a>(PyObject *exception, <b>const</b> <b>char</b> *format, ...)
<a name='L601'><font color='red'>{</font>
<a name='L602'>    va_list vargs;
<a name='L603'>    <a href='../D/3593.html' title='Multiple defined in 2 places.'>PyObject</a>* <a href='../S/500.html#L5' title='Defined at 5 in Include/asdl.h.'>string</a>;
<a name='L604'>
<a name='L605'><font color='darkred'>#ifdef</font> <a href='../D/1566.html' title='Multiple defined in 3 places.'>HAVE_STDARG_PROTOTYPES</a>
<a name='L606'>    va_start(vargs, format);
<a name='L607'><font color='darkred'>#else</font>
<a name='L608'>    va_start(vargs);
<a name='L609'><font color='darkred'>#endif</font>
<a name='L610'>
<a name='L611'>    <a href='../S/500.html#L5' title='Defined at 5 in Include/asdl.h.'>string</a> = <a href='../D/4009.html' title='Multiple defined in 3 places.'>PyUnicode_FromFormatV</a>(format, vargs);
<a name='L612'>    <a href='../S/3106.html#L53' title='Defined at 53 in Python/errors.c.'>PyErr_SetObject</a>(exception, <a href='../S/500.html#L5' title='Defined at 5 in Include/asdl.h.'>string</a>);
<a name='L613'>    <a href='../S/544.html#L767' title='Defined at 767 in Include/object.h.'>Py_XDECREF</a>(<a href='../S/500.html#L5' title='Defined at 5 in Include/asdl.h.'>string</a>);
<a name='L614'>    va_end(vargs);
<a name='L615'>    <b>return</b> <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>;
<a name='L616'><font color='red'>}</font>
<a name='L617'>
<a name='L618'>
<a name='L619'>
<a name='L620'><a href='../D/3593.html' title='Multiple defined in 2 places.'>PyObject</a> *
<a name='L621'><a href='../R/2873.html' title='Multiple refered from 50 places.'>PyErr_NewException</a>(<b>const</b> <b>char</b> *name, PyObject *base, PyObject *dict)
<a name='L622'><font color='red'>{</font>
<a name='L623'>    <b>const</b> <b>char</b> *dot;
<a name='L624'>    <a href='../D/3593.html' title='Multiple defined in 2 places.'>PyObject</a> *modulename = <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>;
<a name='L625'>    <a href='../D/3593.html' title='Multiple defined in 2 places.'>PyObject</a> *classname = <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>;
<a name='L626'>    <a href='../D/3593.html' title='Multiple defined in 2 places.'>PyObject</a> *mydict = <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>;
<a name='L627'>    <a href='../D/3593.html' title='Multiple defined in 2 places.'>PyObject</a> *bases = <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>;
<a name='L628'>    <a href='../D/3593.html' title='Multiple defined in 2 places.'>PyObject</a> *<a href='../S/2250.html#L579' title='Defined at 579 in Modules/_ctypes/callproc.c.'>result</a> = <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>;
<a name='L629'>    dot = strrchr(name, '.');
<a name='L630'>    <b>if</b> (dot == <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>) <font color='red'>{</font>
<a name='L631'>        <a href='../S/3106.html#L122' title='Defined at 122 in Python/errors.c.'>PyErr_SetString</a>(PyExc_SystemError,
<a name='L632'>            "PyErr_NewException: name must be module.class");
<a name='L633'>        <b>return</b> <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>;
<a name='L634'>    <font color='red'>}</font>
<a name='L635'>    <b>if</b> (base == <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>)
<a name='L636'>        base = PyExc_Exception;
<a name='L637'>    <b>if</b> (dict == <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>) <font color='red'>{</font>
<a name='L638'>        dict = mydict = <a href='../S/2787.html#L233' title='Defined at 233 in Objects/dictobject.c.'>PyDict_New</a>();
<a name='L639'>        <b>if</b> (dict == <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>)
<a name='L640'>            <b>goto</b> failure;
<a name='L641'>    <font color='red'>}</font>
<a name='L642'>    <b>if</b> (<a href='../S/2787.html#L2168' title='Defined at 2168 in Objects/dictobject.c.'>PyDict_GetItemString</a>(dict, "__module__") == <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>) <font color='red'>{</font>
<a name='L643'>        modulename = <a href='../D/4013.html' title='Multiple defined in 3 places.'>PyUnicode_FromStringAndSize</a>(name,
<a name='L644'>                                             (<a href='../D/4364.html' title='Multiple defined in 6 places.'>Py_ssize_t</a>)(dot-name));
<a name='L645'>        <b>if</b> (modulename == <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>)
<a name='L646'>            <b>goto</b> failure;
<a name='L647'>        <b>if</b> (<a href='../S/2787.html#L2180' title='Defined at 2180 in Objects/dictobject.c.'>PyDict_SetItemString</a>(dict, "__module__", modulename) != 0)
<a name='L648'>            <b>goto</b> failure;
<a name='L649'>    <font color='red'>}</font>
<a name='L650'>    <b>if</b> (PyTuple_Check(base)) <font color='red'>{</font>
<a name='L651'>        bases = base;
<a name='L652'>        <i><font color='green'>/* INCREF as we create a new ref in the else branch */</font></i>
<a name='L653'>        Py_INCREF(bases);
<a name='L654'>    <font color='red'>}</font> <b>else</b> <font color='red'>{</font>
<a name='L655'>        bases = <a href='../S/2823.html#L186' title='Defined at 186 in Objects/tupleobject.c.'>PyTuple_Pack</a>(1, base);
<a name='L656'>        <b>if</b> (bases == <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>)
<a name='L657'>            <b>goto</b> failure;
<a name='L658'>    <font color='red'>}</font>
<a name='L659'>    <i><font color='green'>/* Create a real new-style class. */</font></i>
<a name='L660'>    <a href='../S/2250.html#L579' title='Defined at 579 in Modules/_ctypes/callproc.c.'>result</a> = <a href='../D/3601.html' title='Multiple defined in 2 places.'>PyObject_CallFunction</a>((<a href='../D/3593.html' title='Multiple defined in 2 places.'>PyObject</a> *)&amp;PyType_Type, "sOO",
<a name='L661'>                                   dot+1, bases, dict);
<a name='L662'>  failure:
<a name='L663'>    <a href='../S/544.html#L767' title='Defined at 767 in Include/object.h.'>Py_XDECREF</a>(bases);
<a name='L664'>    <a href='../S/544.html#L767' title='Defined at 767 in Include/object.h.'>Py_XDECREF</a>(mydict);
<a name='L665'>    <a href='../S/544.html#L767' title='Defined at 767 in Include/object.h.'>Py_XDECREF</a>(classname);
<a name='L666'>    <a href='../S/544.html#L767' title='Defined at 767 in Include/object.h.'>Py_XDECREF</a>(modulename);
<a name='L667'>    <b>return</b> <a href='../S/2250.html#L579' title='Defined at 579 in Modules/_ctypes/callproc.c.'>result</a>;
<a name='L668'><font color='red'>}</font>
<a name='L669'>
<a name='L670'>
<a name='L671'><i><font color='green'>/* Create an exception with docstring */</font></i>
<a name='L672'><a href='../D/3593.html' title='Multiple defined in 2 places.'>PyObject</a> *
<a name='L673'><a href='../S/2614.html#L2231' title='Refered from 2231 in Modules/_testcapimodule.c.'>PyErr_NewExceptionWithDoc</a>(<b>const</b> <b>char</b> *name, <b>const</b> <b>char</b> *doc,
<a name='L674'>                          PyObject *base, PyObject *dict)
<a name='L675'><font color='red'>{</font>
<a name='L676'>    <b>int</b> <a href='../S/2250.html#L579' title='Defined at 579 in Modules/_ctypes/callproc.c.'>result</a>;
<a name='L677'>    <a href='../D/3593.html' title='Multiple defined in 2 places.'>PyObject</a> *ret = <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>;
<a name='L678'>    <a href='../D/3593.html' title='Multiple defined in 2 places.'>PyObject</a> *mydict = <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>; <i><font color='green'>/* points to the dict only if we create it */</font></i>
<a name='L679'>    <a href='../D/3593.html' title='Multiple defined in 2 places.'>PyObject</a> *docobj;
<a name='L680'>
<a name='L681'>    <b>if</b> (dict == <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>) <font color='red'>{</font>
<a name='L682'>        dict = mydict = <a href='../S/2787.html#L233' title='Defined at 233 in Objects/dictobject.c.'>PyDict_New</a>();
<a name='L683'>        <b>if</b> (dict == <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>) <font color='red'>{</font>
<a name='L684'>            <b>return</b> <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>;
<a name='L685'>        <font color='red'>}</font>
<a name='L686'>    <font color='red'>}</font>
<a name='L687'>
<a name='L688'>    <b>if</b> (doc != <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>) <font color='red'>{</font>
<a name='L689'>        docobj = <a href='../D/4012.html' title='Multiple defined in 3 places.'>PyUnicode_FromString</a>(doc);
<a name='L690'>        <b>if</b> (docobj == <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>)
<a name='L691'>            <b>goto</b> failure;
<a name='L692'>        <a href='../S/2250.html#L579' title='Defined at 579 in Modules/_ctypes/callproc.c.'>result</a> = <a href='../S/2787.html#L2180' title='Defined at 2180 in Objects/dictobject.c.'>PyDict_SetItemString</a>(dict, "__doc__", docobj);
<a name='L693'>        Py_DECREF(docobj);
<a name='L694'>        <b>if</b> (<a href='../S/2250.html#L579' title='Defined at 579 in Modules/_ctypes/callproc.c.'>result</a> &lt; 0)
<a name='L695'>            <b>goto</b> failure;
<a name='L696'>    <font color='red'>}</font>
<a name='L697'>
<a name='L698'>    ret = <a href='../S/3106.html#L621' title='Defined at 621 in Python/errors.c.'>PyErr_NewException</a>(name, base, dict);
<a name='L699'>  failure:
<a name='L700'>    <a href='../S/544.html#L767' title='Defined at 767 in Include/object.h.'>Py_XDECREF</a>(mydict);
<a name='L701'>    <b>return</b> ret;
<a name='L702'><font color='red'>}</font>
<a name='L703'>
<a name='L704'>
<a name='L705'><i><font color='green'>/* Call when an exception has occurred but there is no way for Python</font></i>
<a name='L706'><i><font color='green'>   to handle it.  Examples: exception in __del__ or during GC. */</font></i>
<a name='L707'><b>void</b>
<a name='L708'><a href='../R/2901.html' title='Multiple refered from 27 places.'>PyErr_WriteUnraisable</a>(PyObject *obj)
<a name='L709'><font color='red'>{</font>
<a name='L710'>    <a href='../D/3593.html' title='Multiple defined in 2 places.'>PyObject</a> *f, *t, *v, *tb;
<a name='L711'>    <a href='../S/3106.html#L304' title='Defined at 304 in Python/errors.c.'>PyErr_Fetch</a>(&amp;t, &amp;v, &amp;tb);
<a name='L712'>    f = <a href='../S/3144.html#L44' title='Defined at 44 in Python/sysmodule.c.'>PySys_GetObject</a>("stderr");
<a name='L713'>    <b>if</b> (f != <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a> &amp;&amp; f != Py_None) <font color='red'>{</font>
<a name='L714'>        <a href='../S/2790.html#L167' title='Defined at 167 in Objects/fileobject.c.'>PyFile_WriteString</a>("Exception ", f);
<a name='L715'>        <b>if</b> (t) <font color='red'>{</font>
<a name='L716'>            <a href='../D/3593.html' title='Multiple defined in 2 places.'>PyObject</a>* moduleName;
<a name='L717'>            <b>char</b>* className;
<a name='L718'>            <a href='../D/6814.html' title='Multiple defined in 2 places.'>assert</a>(PyExceptionClass_Check(t));
<a name='L719'>            className = PyExceptionClass_Name(t);
<a name='L720'>            <b>if</b> (className != <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>) <font color='red'>{</font>
<a name='L721'>                <b>char</b> *dot = strrchr(className, '.');
<a name='L722'>                <b>if</b> (dot != <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>)
<a name='L723'>                    className = dot+1;
<a name='L724'>            <font color='red'>}</font>
<a name='L725'>
<a name='L726'>            moduleName = <a href='../S/2803.html#L779' title='Defined at 779 in Objects/object.c.'>PyObject_GetAttrString</a>(t, "__module__");
<a name='L727'>            <b>if</b> (moduleName == <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>)
<a name='L728'>                <a href='../S/2790.html#L167' title='Defined at 167 in Objects/fileobject.c.'>PyFile_WriteString</a>("&lt;unknown&gt;", f);
<a name='L729'>            <b>else</b> <font color='red'>{</font>
<a name='L730'>                <b>char</b>* modstr = <a href='../S/2828.html#L1919' title='Defined at 1919 in Objects/unicodeobject.c.'>_PyUnicode_AsString</a>(moduleName);
<a name='L731'>                <b>if</b> (modstr &amp;&amp;
<a name='L732'>                    strcmp(modstr, "builtins") != 0)
<a name='L733'>                <font color='red'>{</font>
<a name='L734'>                    <a href='../S/2790.html#L167' title='Defined at 167 in Objects/fileobject.c.'>PyFile_WriteString</a>(modstr, f);
<a name='L735'>                    <a href='../S/2790.html#L167' title='Defined at 167 in Objects/fileobject.c.'>PyFile_WriteString</a>(".", f);
<a name='L736'>                <font color='red'>}</font>
<a name='L737'>            <font color='red'>}</font>
<a name='L738'>            <b>if</b> (className == <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>)
<a name='L739'>                <a href='../S/2790.html#L167' title='Defined at 167 in Objects/fileobject.c.'>PyFile_WriteString</a>("&lt;unknown&gt;", f);
<a name='L740'>            <b>else</b>
<a name='L741'>                <a href='../S/2790.html#L167' title='Defined at 167 in Objects/fileobject.c.'>PyFile_WriteString</a>(className, f);
<a name='L742'>            <b>if</b> (v &amp;&amp; v != Py_None) <font color='red'>{</font>
<a name='L743'>                <a href='../S/2790.html#L167' title='Defined at 167 in Objects/fileobject.c.'>PyFile_WriteString</a>(": ", f);
<a name='L744'>                <a href='../S/2790.html#L131' title='Defined at 131 in Objects/fileobject.c.'>PyFile_WriteObject</a>(v, f, 0);
<a name='L745'>            <font color='red'>}</font>
<a name='L746'>            <a href='../S/544.html#L767' title='Defined at 767 in Include/object.h.'>Py_XDECREF</a>(moduleName);
<a name='L747'>        <font color='red'>}</font>
<a name='L748'>        <b>if</b> (obj) <font color='red'>{</font>
<a name='L749'>            <a href='../S/2790.html#L167' title='Defined at 167 in Objects/fileobject.c.'>PyFile_WriteString</a>(" in ", f);
<a name='L750'>            <a href='../S/2790.html#L131' title='Defined at 131 in Objects/fileobject.c.'>PyFile_WriteObject</a>(obj, f, 0);
<a name='L751'>        <font color='red'>}</font>
<a name='L752'>        <a href='../S/2790.html#L167' title='Defined at 167 in Objects/fileobject.c.'>PyFile_WriteString</a>(" ignored\n", f);
<a name='L753'>        <a href='../S/3106.html#L318' title='Defined at 318 in Python/errors.c.'>PyErr_Clear</a>(); <i><font color='green'>/* Just in case */</font></i>
<a name='L754'>    <font color='red'>}</font>
<a name='L755'>    <a href='../S/544.html#L767' title='Defined at 767 in Include/object.h.'>Py_XDECREF</a>(t);
<a name='L756'>    <a href='../S/544.html#L767' title='Defined at 767 in Include/object.h.'>Py_XDECREF</a>(v);
<a name='L757'>    <a href='../S/544.html#L767' title='Defined at 767 in Include/object.h.'>Py_XDECREF</a>(tb);
<a name='L758'><font color='red'>}</font>
<a name='L759'>
<a name='L760'><b>extern</b> <a href='../D/3593.html' title='Multiple defined in 2 places.'>PyObject</a> *<a href='../S/3138.html#L96' title='Defined at 96 in Python/pythonrun.c.'>PyModule_GetWarningsModule</a>(<b>void</b>);
<a name='L761'>
<a name='L762'>
<a name='L763'><b>void</b>
<a name='L764'>PyErr_SyntaxLocation(<b>const</b> <b>char</b> *filename, <b>int</b> lineno) <font color='red'>{</font>
<a name='L765'>    <a href='../S/3106.html#L774' title='Defined at 774 in Python/errors.c.'>PyErr_SyntaxLocationEx</a>(filename, lineno, -1);
<a name='L766'><font color='red'>}</font>
<a name='L767'>
<a name='L768'>
<a name='L769'><i><font color='green'>/* Set file and line information for the current exception.</font></i>
<a name='L770'><i><font color='green'>   If the exception is not a SyntaxError, also sets additional attributes</font></i>
<a name='L771'><i><font color='green'>   to make printing of exceptions believe it is a syntax error. */</font></i>
<a name='L772'>
<a name='L773'><b>void</b>
<a name='L774'><a href='../R/2897.html' title='Multiple refered from 11 places.'>PyErr_SyntaxLocationEx</a>(<b>const</b> <b>char</b> *filename, <b>int</b> lineno, <b>int</b> col_offset)
<a name='L775'><font color='red'>{</font>
<a name='L776'>    <a href='../D/3593.html' title='Multiple defined in 2 places.'>PyObject</a> *exc, *v, *tb, *tmp;
<a name='L777'>
<a name='L778'>    <i><font color='green'>/* add attributes for the line number and filename for the error */</font></i>
<a name='L779'>    <a href='../S/3106.html#L304' title='Defined at 304 in Python/errors.c.'>PyErr_Fetch</a>(&amp;exc, &amp;v, &amp;tb);
<a name='L780'>    <a href='../S/3106.html#L202' title='Defined at 202 in Python/errors.c.'>PyErr_NormalizeException</a>(&amp;exc, &amp;v, &amp;tb);
<a name='L781'>    <i><font color='green'>/* XXX check that it is, indeed, a syntax error. It might not</font></i>
<a name='L782'><i><font color='green'>     * be, though. */</font></i>
<a name='L783'>    tmp = <a href='../S/2799.html#L176' title='Defined at 176 in Objects/longobject.c.'>PyLong_FromLong</a>(lineno);
<a name='L784'>    <b>if</b> (tmp == <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>)
<a name='L785'>        <a href='../S/3106.html#L318' title='Defined at 318 in Python/errors.c.'>PyErr_Clear</a>();
<a name='L786'>    <b>else</b> <font color='red'>{</font>
<a name='L787'>        <b>if</b> (<a href='../S/2803.html#L806' title='Defined at 806 in Objects/object.c.'>PyObject_SetAttrString</a>(v, "lineno", tmp))
<a name='L788'>            <a href='../S/3106.html#L318' title='Defined at 318 in Python/errors.c.'>PyErr_Clear</a>();
<a name='L789'>        Py_DECREF(tmp);
<a name='L790'>    <font color='red'>}</font>
<a name='L791'>    <b>if</b> (col_offset &gt;= 0) <font color='red'>{</font>
<a name='L792'>        tmp = <a href='../S/2799.html#L176' title='Defined at 176 in Objects/longobject.c.'>PyLong_FromLong</a>(col_offset);
<a name='L793'>        <b>if</b> (tmp == <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>)
<a name='L794'>            <a href='../S/3106.html#L318' title='Defined at 318 in Python/errors.c.'>PyErr_Clear</a>();
<a name='L795'>        <b>else</b> <font color='red'>{</font>
<a name='L796'>            <b>if</b> (<a href='../S/2803.html#L806' title='Defined at 806 in Objects/object.c.'>PyObject_SetAttrString</a>(v, "offset", tmp))
<a name='L797'>                <a href='../S/3106.html#L318' title='Defined at 318 in Python/errors.c.'>PyErr_Clear</a>();
<a name='L798'>            Py_DECREF(tmp);
<a name='L799'>        <font color='red'>}</font>
<a name='L800'>    <font color='red'>}</font>
<a name='L801'>    <b>if</b> (filename != <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>) <font color='red'>{</font>
<a name='L802'>        tmp = <a href='../D/3975.html' title='Multiple defined in 3 places.'>PyUnicode_DecodeFSDefault</a>(filename);
<a name='L803'>        <b>if</b> (tmp == <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>)
<a name='L804'>            <a href='../S/3106.html#L318' title='Defined at 318 in Python/errors.c.'>PyErr_Clear</a>();
<a name='L805'>        <b>else</b> <font color='red'>{</font>
<a name='L806'>            <b>if</b> (<a href='../S/2803.html#L806' title='Defined at 806 in Objects/object.c.'>PyObject_SetAttrString</a>(v, "filename", tmp))
<a name='L807'>                <a href='../S/3106.html#L318' title='Defined at 318 in Python/errors.c.'>PyErr_Clear</a>();
<a name='L808'>            Py_DECREF(tmp);
<a name='L809'>        <font color='red'>}</font>
<a name='L810'>
<a name='L811'>        tmp = <a href='../S/3106.html#L848' title='Defined at 848 in Python/errors.c.'>PyErr_ProgramText</a>(filename, lineno);
<a name='L812'>        <b>if</b> (tmp) <font color='red'>{</font>
<a name='L813'>            <b>if</b> (<a href='../S/2803.html#L806' title='Defined at 806 in Objects/object.c.'>PyObject_SetAttrString</a>(v, "text", tmp))
<a name='L814'>                <a href='../S/3106.html#L318' title='Defined at 318 in Python/errors.c.'>PyErr_Clear</a>();
<a name='L815'>            Py_DECREF(tmp);
<a name='L816'>        <font color='red'>}</font>
<a name='L817'>    <font color='red'>}</font>
<a name='L818'>    <b>if</b> (<a href='../S/2803.html#L806' title='Defined at 806 in Objects/object.c.'>PyObject_SetAttrString</a>(v, "offset", Py_None)) <font color='red'>{</font>
<a name='L819'>        <a href='../S/3106.html#L318' title='Defined at 318 in Python/errors.c.'>PyErr_Clear</a>();
<a name='L820'>    <font color='red'>}</font>
<a name='L821'>    <b>if</b> (exc != PyExc_SyntaxError) <font color='red'>{</font>
<a name='L822'>        <b>if</b> (!<a href='../S/2803.html#L794' title='Defined at 794 in Objects/object.c.'>PyObject_HasAttrString</a>(v, "msg")) <font color='red'>{</font>
<a name='L823'>            tmp = <a href='../S/2803.html#L396' title='Defined at 396 in Objects/object.c.'>PyObject_Str</a>(v);
<a name='L824'>            <b>if</b> (tmp) <font color='red'>{</font>
<a name='L825'>                <b>if</b> (<a href='../S/2803.html#L806' title='Defined at 806 in Objects/object.c.'>PyObject_SetAttrString</a>(v, "msg", tmp))
<a name='L826'>                    <a href='../S/3106.html#L318' title='Defined at 318 in Python/errors.c.'>PyErr_Clear</a>();
<a name='L827'>                Py_DECREF(tmp);
<a name='L828'>            <font color='red'>}</font> <b>else</b> <font color='red'>{</font>
<a name='L829'>                <a href='../S/3106.html#L318' title='Defined at 318 in Python/errors.c.'>PyErr_Clear</a>();
<a name='L830'>            <font color='red'>}</font>
<a name='L831'>        <font color='red'>}</font>
<a name='L832'>        <b>if</b> (!<a href='../S/2803.html#L794' title='Defined at 794 in Objects/object.c.'>PyObject_HasAttrString</a>(v, "print_file_and_line")) <font color='red'>{</font>
<a name='L833'>            <b>if</b> (<a href='../S/2803.html#L806' title='Defined at 806 in Objects/object.c.'>PyObject_SetAttrString</a>(v, "print_file_and_line",
<a name='L834'>                                       Py_None))
<a name='L835'>                <a href='../S/3106.html#L318' title='Defined at 318 in Python/errors.c.'>PyErr_Clear</a>();
<a name='L836'>        <font color='red'>}</font>
<a name='L837'>    <font color='red'>}</font>
<a name='L838'>    <a href='../S/3106.html#L25' title='Defined at 25 in Python/errors.c.'>PyErr_Restore</a>(exc, v, tb);
<a name='L839'><font color='red'>}</font>
<a name='L840'>
<a name='L841'><i><font color='green'>/* Attempt to load the line of text that the exception refers to.  If it</font></i>
<a name='L842'><i><font color='green'>   fails, it will return NULL but will not set an exception.</font></i>
<a name='L843'><i><font color='green'></font></i>
<a name='L844'><i><font color='green'>   XXX The functionality of this function is quite similar to the</font></i>
<a name='L845'><i><font color='green'>   functionality in tb_displayline() in traceback.c. */</font></i>
<a name='L846'>
<a name='L847'><a href='../D/3593.html' title='Multiple defined in 2 places.'>PyObject</a> *
<a name='L848'><a href='../R/2880.html' title='Multiple refered from 3 places.'>PyErr_ProgramText</a>(<b>const</b> <b>char</b> *filename, <b>int</b> lineno)
<a name='L849'><font color='red'>{</font>
<a name='L850'>    FILE *fp;
<a name='L851'>    <b>int</b> i;
<a name='L852'>    <b>char</b> linebuf[1000];
<a name='L853'>
<a name='L854'>    <b>if</b> (filename == <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a> || *filename == '\0' || lineno &lt;= 0)
<a name='L855'>        <b>return</b> <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>;
<a name='L856'>    fp = fopen(filename, "r" <a href='../S/523.html#L9' title='Defined at 9 in Include/fileobject.h.'>PY_STDIOTEXTMODE</a>);
<a name='L857'>    <b>if</b> (fp == <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>)
<a name='L858'>        <b>return</b> <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>;
<a name='L859'>    <b>for</b> (i = 0; i &lt; lineno; i++) <font color='red'>{</font>
<a name='L860'>        <b>char</b> *pLastChar = &amp;linebuf[<b>sizeof</b>(linebuf) - 2];
<a name='L861'>        <b>do</b> <font color='red'>{</font>
<a name='L862'>            *pLastChar = '\0';
<a name='L863'>            <b>if</b> (<a href='../D/4299.html' title='Multiple defined in 2 places.'>Py_UniversalNewlineFgets</a>(linebuf, <b>sizeof</b> linebuf,
<a name='L864'>                                         fp, <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>) == <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>)
<a name='L865'>                <b>break</b>;
<a name='L866'>            <i><font color='green'>/* fgets read *something*; if it didn't get as</font></i>
<a name='L867'><i><font color='green'>               far as pLastChar, it must have found a newline</font></i>
<a name='L868'><i><font color='green'>               or hit the end of the file; if pLastChar is \n,</font></i>
<a name='L869'><i><font color='green'>               it obviously found a newline; else we haven't</font></i>
<a name='L870'><i><font color='green'>               yet seen a newline, so must continue */</font></i>
<a name='L871'>        <font color='red'>}</font> <b>while</b> (*pLastChar != '\0' &amp;&amp; *pLastChar != '\n');
<a name='L872'>    <font color='red'>}</font>
<a name='L873'>    fclose(fp);
<a name='L874'>    <b>if</b> (i == lineno) <font color='red'>{</font>
<a name='L875'>        <b>char</b> *p = linebuf;
<a name='L876'>        <a href='../D/3593.html' title='Multiple defined in 2 places.'>PyObject</a> *res;
<a name='L877'>        <b>while</b> (*p == ' ' || *p == '\t' || *p == '\014')
<a name='L878'>            p++;
<a name='L879'>        res = <a href='../D/4012.html' title='Multiple defined in 3 places.'>PyUnicode_FromString</a>(p);
<a name='L880'>        <b>if</b> (res == <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>)
<a name='L881'>            <a href='../S/3106.html#L318' title='Defined at 318 in Python/errors.c.'>PyErr_Clear</a>();
<a name='L882'>        <b>return</b> res;
<a name='L883'>    <font color='red'>}</font>
<a name='L884'>    <b>return</b> <a href='../S/2891.html#L5' title='Defined at 5 in PC/os2emx/dllentry.c.'>NULL</a>;
<a name='L885'><font color='red'>}</font>
<a name='L886'>
<a name='L887'><font color='darkred'>#ifdef</font> __cplusplus
<a name='L888'><font color='red'>}</font>
<a name='L889'><font color='darkred'>#endif</font>
</pre>
<hr>
<a name='BOTTOM'>
<i><font color='green'>/* [&lt;][&gt;]<a href='#L25'>[^]</a><a href='#L848'>[v]</a><a href='#TOP'>[top]</a>[bottom]<a href='../mains.html'>[index]</a><a href='../help.html'>[help]</a> */</font></i>
</body>
</html>
